onnx/defs/training/defs.cc

```diff
@@ -0,0 +1,171 @@
+```
+
+, where the content inside the braces of Gradient are attributes. The attribute
+"xs" specifies the necessary inputs to compute the variable specified by
+attribute "y". To compute gradient, the runtime should identify a sub-graph with
+inputs specified by "xs" and outputs containing "y"'s content and do a backward
+pass (or other auto-differentiation techniques whenever needed).
+
+Since X is not trainable tensor, the user can avoid the creation of dY/dX by
+assigning an empty string to the 1st output name of that Gradient. In other
+words, all Gradient's outputs are optional. Note that the concept of optional
+outputs can also be found in ONNX's RNN, GRU, and LSTM.
+
+Gradient operator can compute derivative against intermediate tensors. For
+example, the gradient of Y with respect to H can be done in
+
+```
+W --> Conv --> H --> Gemm --> Y
+       ^       |      ^
+       |       |      |
+       X       |      Z
+       .-------'      |
+       |   .----------'
+       |   | (H/Z is the 1st/2nd input of Gradient as shown in "xs")
+       v   v
+      Gradient(xs=["H", "Z"], y=["Y"])
+       |   |
+       |   '-----------------------------------> dY/dH (1st output of Gradient)
+       |
+       '---------------------------------------> dY/dZ (2nd output of Gradient)
+```
+
+It is possible to represent high-order differentiation using Gradient operator.
+An example for linear model is
+
+```
+W --> Gemm --> Y --> Loss --> O
+       ^              ^
+       |              |
+       X              L
+```
+
+To compute the 2nd order derivative of O with respect to W (denoted by
+d^2O/dW^2), one can do
+
+```
+W --> Gemm --> Y --> Loss --> O
+|      ^              ^
+|      |              |
+|      X              L
+|      |
+|      |
++------+--> Gradient(xs=["X", "W"], y=["Y"]) ---> dO/dX (1st output of Gradient)
+|      |      |
+|      |      '---> dO/dW (2nd output of Gradient)
+|      v
+'---> Gradient(xs=["X", "W"], y=["dO/dW"]) ---> d(dO/dW)dX (1st output of
+       |                                          Gradient)
+       |
+       |
+       '---> d^2O/dW^2 (2nd output of Gradient)
+```
+
+As mentioned above, the attributes "xs" and "y" are used to identify a graph,
+and we can feed different tensors to the identified graph. For,
+example, one can compute the gradient of Y with respect to H by substituting
+Y_1 into Y and H_1 into H.
+
+```
+W --> Conv --> H --> Gemm --> Y
+       ^              ^
+       |              |
+       X              Z
+
+           Z_1 (the 2nd input of Gradient)
+           |
+           v
+W_1 --> Gradient(xs=["H", "Z"], y=["Y"]) ---> dY/dX when Y = Y_1
+         |   |
+         |   '-----------------------------------> dY/dW (2nd output of Gradient)
+         |
+         '---------------------------------------> dY/dZ (3rd output of Gradient)
+```
+
+This graph also implies that one can identify a gradient graph once and reuse
+it multiple times with different input values.
```

Comment:
How? We don't have a way to give a name to the gradient-graph and reuse it. Do you mean that we can have multiple gradient nodes with the same attribute values?

Reply:
If you put `Gradient` in a loop, it can receive different values from loop-carried dependencies.